File: NBA_taehak/2023_03/week_1st/43105.py
# [Programmers] 43105. 정수 삼각형
# 소요 시간 : 00분

def solution(triangle):
    answer = 0

    for row in range(1, len(triangle)):
        for idx in range(len(triangle[row])):
            if idx == 0:
                triangle[row][idx] += triangle[row-1][0]
            elif idx == len(triangle[row]) - 1:
                triangle[row][idx] += triangle[row-1][-1]
            else:
                triangle[row][idx] += max(triangle[row-1][idx-1], triangle[row-1][idx])

    answer = max(triangle[-1])

    return answer

File: NBA_deokgyun/2023_03/week_1st/43105.java
// [Programmers] 43105. 정수 삼각형
// 소요 시간 : 60분

import java.util.*;
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int tri_len = triangle.length;
        int[][] num_arr = triangle;
        for (int i = 0; i < tri_len; i++) {
            int arr_len = triangle[i].length;
            for (int j = 0; j < arr_len;j++) {
                if (i == 0) {
                    num_arr[i][j] = triangle[i][j];
                }
                else if (j == 0) {
                    num_arr[i][j] = triangle[i - 1][j] + triangle[i][j];
                }
                else if (j == arr_len - 1) {
                    num_arr[i][j] = triangle[i - 1][j - 1] + triangle[i][j];
                }
                else {
                    int num1 = triangle[i - 1][j - 1] + triangle[i][j];
                    int num2 = triangle[i - 1][j] + triangle[i][j];
                    num_arr[i][j]  = (num1 > num2) ? num1 : num2;
                }
            }
        }
        Arrays.sort(num_arr[tri_len - 1]);
        answer = num_arr[tri_len - 1][num_arr[tri_len - 1].length - 1];
        return answer;
    }
}